src/pybind/mgr/cephadm/agent.py

```diff
@@ -130,22 +146,112 @@ def shutdown(self) -> None:
         self.cherrypy_shutdown_event.set()
 
 
-class Root:
-    exposed = True
+class Root(object):
+
+    # collapse everything to '/'
+    def _cp_dispatch(self, vpath: str) -> 'Root':
+        cherrypy.request.path = ''
+        return self
 
     def __init__(self, mgr: "CephadmOrchestrator"):
         self.mgr = mgr
-        self.data = HostData(self.mgr)
 
-    def GET(self) -> str:
+    @cherrypy.expose
+    def index(self) -> str:
         return '''<!DOCTYPE html>
 <html>
 <head><title>Cephadm HTTP Endpoint</title></head>
 <body>
-<p>Cephadm HTTP Endpoint is up and running</p>
+<p>Cephadm Service discovery Endpoints</p>
+<p><a href='sd-config?service=prometheus'>Prometheus http sd-config</a></p>
+<p><a href='sd-config?service=alert-manager'>Prometheus http sd-config</a></p>
+<p><a href='sd-config?service=node-exporter'>Node exporter http sd-config</a></p>
+<p><a href='sd-config?service=haproxy'>HAProxy http sd-config</a></p>
+<p><a href='rules'>Prometheus rules</a></p>
 </body>
 </html>'''
 
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def get_sd_config(self, service: str) -> List[Dict[str, Collection[str]]]:
+        if service == 'prometheus':
+            return self.prometheus_sd_config()
+        elif service == 'alert-manager':
+            return self.alertmgr_sd_config()
+        elif service == 'node-exporter':
+            return self.node_exporter_sd_config()
+        elif service == 'haproxy':
+            return self.haproxy_sd_config()
+        else:
+            return [{'error': {f"Unknown service type '{service}'"}}]
```

Comment:
When I used prometheus/sd_config?service=bobbins I get a http 500 and the cephadm shows a traceback.
```
2022-03-23T04:41:01.806629+0000 mgr.ceph-mstr-aio.rthnhs (mgr.14170) 1221 : cephadm [INF] Detected new or changed devices on ceph-mstr-aio
2022-03-23T04:41:19.200652+0000 mgr.ceph-mstr-aio.rthnhs (mgr.14170) 1231 : cephadm [ERR] [23/Mar/2022:04:41:19] HTTP 
Traceback (most recent call last):
  File "/lib/python3.6/site-packages/cherrypy/_cprequest.py", line 638, in respond
    self._do_respond(path_info)
  File "/lib/python3.6/site-packages/cherrypy/_cprequest.py", line 702, in _do_respond
    response.finalize()
  File "/lib/python3.6/site-packages/cherrypy/_cprequest.py", line 898, in finalize
    content = self.collapse_body()
  File "/lib/python3.6/site-packages/cherrypy/_cprequest.py", line 856, in collapse_body
    new_body = b''.join(self.body)
  File "/lib/python3.6/site-packages/cherrypy/_json.py", line 24, in encode
    for chunk in _encode(value):
  File "/lib64/python3.6/json/encoder.py", line 428, in _iterencode
    yield from _iterencode_list(o, _current_indent_level)
  File "/lib64/python3.6/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/lib64/python3.6/json/encoder.py", line 404, in _iterencode_dict
    yield from chunks
  File "/lib64/python3.6/json/encoder.py", line 437, in _iterencode
    o = _default(o)
  File "/lib64/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'set' is not JSON serializable
```

Reply:
prometheus expects a 200 - so I suspect for an invalid group we should be returning an empty list and a 200.

Reply:
Fixed. There was an issue with the returned value.

Reply:
Great